api/envoy/extensions/filters/common/set_filter_state/v3/rule.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.common.set_filter_state.v3;
+
+import "envoy/config/core/v3/substitution_format_string.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.common.set_filter_state.v3";
+option java_outer_classname = "RuleProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/common/set_filter_state/v3;set_filter_statev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Set-Filter-State Update Rule]
+
+// A filter state update rule.
+// [#next-free-field: 6]
+message Rule {
+  enum SharedWithUpstream {
+    // Object is not shared with the upstream internal connections.
+    NONE = 0;
+
+    // Object is shared with the upstream internal connection.
+    ONCE = 1;
+
+    // Object is shared with the upstream internal connection and any internal connection upstream from it.
+    TRANSITIVE = 2;
+  }
+
+  // Filter state key.
+  string key = 1 [(validate.rules).string = {min_len: 1}];
+
+  oneof value {
+    option (validate.required) = true;
+
+    // Uses the :ref:`format string <config_access_log_format_strings>` to
+    // instantiate the filter state object value.
+    config.core.v3.SubstitutionFormatString format_string = 2;
+  }
+
+  // If marked as read-only, the filter state key cannot be overridden by the subsequent updates.
+  bool read_only = 3;
```

Comment:
This reflects the filter state implementation in stream info. It has the ability to lock-on-write, but not much beyond that.

Reply:
Does `subsequent updates` actually imply:
```
... the filter state key cannot be overridden by operations of filters that are executed after the current Rule sets the key's value.
If ``skip_if_empty`` is true and the value is empty, subsequent filter operations may update the value.
```

I still think it may make more sense to conditionally update (similar to [HeaderAppendAction](https://github.com/envoyproxy/envoy/blob/2a0f2657d2af838956daa4e7cd589eb5e925edfd/api/envoy/config/core/v3/base.proto#L348)) but this is orthogonal to this field. I couldn't find a better name that clarifies future-looking read-only, so I guess it can be kept in-sync with the original FilterState naming.

Reply:
No, this is more like "const", the value is assigned once, and cannot be changed by anything. I updated the comment to reflect that.